File: Validation/Performance/scripts/cmsScimarkStop.pl
#!/usr/bin/perl
#Script to stop and log the output of cmsScimark

#Get some environment variables to use
#$CMSSW_BASE=$ENV{'CMSSW_BASE'};
#$CMSSW_RELEASE_BASE=$ENV{'CMSSW_RELEASE_BASE'};

#Adding a check for a local version of the packages
#$PerformancePkg="$CMSSW_BASE/src/Validation/Performance";
#if (-e $PerformancePkg)
#{
#    $BASE_PERFORMANCE=$PerformancePkg;
#    print "**[cmsScimarkStop.pl]Using LOCAL version of Validation/Performance instead of the RELEASE version**\n";
#}
#else
#{
#    $BASE_PERFORMANCE="$CMSSW_RELEASE_BASE/src/Validation/Performance";
#}
#Get the PID of the running cmsScimarkLaunch.csh
$PsOutput=`ps -ef|grep cmsScimarkLaunch`;
print $PsOutput;
@Ps=split('\n',$PsOutput);
#Kill all the cmsScimarkLaunch.csh processes
foreach (@Ps)
{
    @PsTokens=split(' ',$_);
    if (($PsTokens[7] eq "/bin/csh")&&($PsTokens[8] =~ /cmsScimarkLaunch.csh/))
    {
	$kill=`kill -9 $PsTokens[1]`;
	print "kill -9 $PsTokens[1] (was $_)\n";
	$cpucore=$PsTokens[9];
	print "Killed cmsScimarkLaunch.csh on cpu$cpucore\n";
	$ResultsDir="cmsScimarkResults_cpu".$cpucore;
#	print "$ResultsDir\n";
	$ResultsLog="cmsScimark_".$cpucore.".log";
#	print "$ResultsLog\n";
	if (!(-e $ResultsDir))
	{
	    $mkdir=`mkdir $ResultsDir`;
	    print "mkdir $ResultsDir\n";
	}
	#Script assumes that cmsScimarkParser.py is already in the release
        #Otherwise, one needs to add ./ in front of it
	$cmsScimarkParse=`cmsScimarkParser.py -i $ResultsLog -o $ResultsDir`;
	print "cmsScimarkParser.py -i $ResultsLog -o $ResultsDir\n";
    }
    else
    {
	print "There was no matching in the ps -ef results";
    }
}
exit;

File: Validation/Performance/scripts/cmsScimarkStop.py
#! /usr/bin/env python
#Script to
#1-check for cmsScimarkLaunch (infinite loop) scripts
#2-kill them
#3-report their results using cmsScimarkParser.py

import subprocess,os,sys

def main():
    #Use ps -ef to look for cmsScimarkLaunch processes
    ps_stdouterr=subprocess.Popen("ps -efww|grep cmsScimarkLaunch|grep -v grep|grep -v 'sh -c'",shell=True,stdout=subprocess.PIPE, stderr=subprocess.STDOUT).stdout
    if ps_stdouterr:
        ps_lines=ps_stdouterr.readlines()
        #print ps_lines
    if ps_lines:
        for line in ps_lines:
            tokens=line.split()
            #Look up the PID
            PID=tokens[1]
            #Look up the cpu core
            core=tokens[9]
            print "Found process:\n%s"%line[:-1] #to eliminate the extra \n
            #Kill the PID
            print "Killing process with PID %s"%PID
            kill_stdouterr=subprocess.Popen("kill %s"%PID,shell=True,stdout=subprocess.PIPE, stderr=subprocess.STDOUT).stdout.read() 
            print kill_stdouterr
            #Harvest the cmsScimark scores
            #Look for the cmsScimark log:
            if os.path.exists("cmsScimark_%s.log"%core): 
                #Create the results dir
                mkdir_stdouterr=subprocess.Popen("mkdir cmsScimarkResults_cpu%s"%core,shell=True,stdout=subprocess.PIPE, stderr=subprocess.STDOUT).stdout.read()
                print mkdir_stdouterr
                #Execute the harvesting scrip cmsScimarkParser.py (it is in the release)
                harvest_stdouterr=subprocess.Popen("cmsScimarkParser.py -i cmsScimark_%s.log -o cmsScimarkResults_cpu%s"%(core,core),shell=True,stdout=subprocess.PIPE, stderr=subprocess.STDOUT).stdout.read()
                print harvest_stdouterr
            else:
                print "No cmsScimark_%s.log file was found for cpu%s, log might be in another directory!"%(core,core)
    else:
        print "No cmsScimarkLaunch processes found in the ps -ef output"
    return 0

if __name__ == "__main__":
    sys.exit(main())
